### ISSUE TITLE
Why can't implicitly authenticated users be permanent/provided a refresh token?

### ISSUE THREAD

**Issue by a9io** [reddit/reddit#1352]
I don't understand why implicit authentication has such little trust compared to explicit.

Plenty of mobile apps/clients use cookie authentication at the moment (set to be deprecated) and they're a client-only application. But what happens when they switch to Oauth?

The time a user spends on the apps is certainly more than an hour (multiple times daily) and it would be a nightmare to have to make the user sign in again after their token expires.

Of course there is *explicit* authentication, but that is only for servers. What if a client is just a client which is used by the user for the long-term? Storing private keys on an app someone downloads is extremely frowned upon.

There is no refresh token for implicitly authenticated users who desire this kind of behavior. And there are a *lot* of those users.

**Comment by rtheunissen** [comment]
If you save the credentials on the device, the client should be able to check if the token is about to expire, and automatically request another one to replace the old one.

**Comment by a9io** [comment]
You can't though. The whole point of the system only allowing auth to last 1 hour is so re-authentication behind the scenes is impossible. I encourage you to try it yourself.

**Comment by rtheunissen** [comment]
I'm not talking about refresh tokens. What's stopping you from requesting another access token to replace one that is about to expire? 

I am doing this currently and it works well. If you can store the credentials (not the token), you can just request a brand new token.

**Comment by rtheunissen** [comment]
I'm not talking about refresh tokens. What's stopping you from requesting another access token to replace one that is about to expire? 

I am doing this currently and it works well. If you can store the credentials (not the token), you can just request a brand new token.

**Comment by a9io** [comment]
What credentials? You can't request an access token without going through the authentication url AFAIK.

**Comment by rtheunissen** [comment]
You said 'implicit' which doesn't use the redirect url?

**Comment by a9io** [comment]
Implicit uses the redirect url. Script-based auth is the one that uses credentials. https://github.com/reddit/reddit/wiki/OAuth2#authorization

**Comment by rtheunissen** [comment]
@a9io You're absolutely right. I was confusing implicit with app-only. I'm now trying to figure out what the difference between code and implicit flow. The documentation is there but it doesn't spell it out for us.

**Comment by rtheunissen** [comment]
:+1:

**Comment by rtheunissen** [comment]
[This thread on /r/redditdev](https://www.reddit.com/r/redditdev/comments/3bfh8i/authenticating_a_client_nowadays/) seems to describe the frustration pretty well.

**Comment by a9io** [comment]
:) glad I'm not alone in this.

**Comment by DorianGray** [comment]
https://www.reddit.com/r/redditdev/comments/3bfh8i/authenticating_a_client_nowadays/csp6548

**Comment by a9io** [comment]
Looks like that might be the solution, then. [You really need to update your docs](https://www.reddit.com/r/redditdev/comments/3bfh8i/authenticating_a_client_nowadays/csp6vgb).

**Comment by rtheunissen** [comment]
Thanks @DorianGray :sparkles:

**Comment by DorianGray** [comment]
Yes, we do =) Honestly, most of our oauth code needs an overhaul. We're working on priorities for this stuff now. Thanks for bringing it to our attention!

**Comment by a9io** [comment]
Awesome! :+1: :+1:

**Comment by a9io** [comment]
It seems that feature is unfinished. https://github.com/trevorsenior/snoocore/issues/135#issuecomment-117863345 There is no CORS header for that endpoint.

**Comment by ThorArakis** [comment]
There should be CORS headers for the access token endpoint for a successful call. Unfortunately for a failed call the error handler goes through a bunch of crazy pylons logic and thus ends up being handled in a totally different place in code which isn't currently adding cors headers. So in the case of access failure you'll also see a CORS exception. It's not as easy to fix as it sounds, but I'll see what I can do.

**Comment by a9io** [comment]
So if it was a valid call then there would be no error thrown? It seems you probably would need to know the error in case of checking for expired tokens, yeah. Thanks.

**Comment by ThorArakis** [comment]
Correct. If you fix the reason that it's 401'ing the CORS error should go away as well.

**Comment by a9io** [comment]
Thanks, let me know when this is pushed to the servers (if it's anytime soon hehe)

**Comment by a9io** [comment]
@ThorArakis is this feature going to removed altogether then? If so at least answer the initial question then :P

**Comment by ThorArakis** [comment]
If the initial question is "why does implicit have so little trust" the answer is because the OAuth spec intends it this way. Longer answer [here](https://www.reddit.com/r/redditdev/comments/3bfh8i/authenticating_a_client_nowadays/csp9nku).

**Comment by a9io** [comment]
ok, I guess mobile clients all use explicit auth then.

**Comment by ThorArakis** [comment]
Yeah, your right that the auth scenario for 3rd party mobile clients isn't really solved yet.
